python/pyspark/sql/datasource.py

```diff
@@ -357,6 +363,30 @@ def abort(self, messages: List["WriterCommitMessage"]) -> None:
         ...
 
 
+class DataStreamOffset:
+    def __init__(self, value: Any) -> None:
+        self.value = value
+
+    def set_value(self, json_str: str) -> None:
+        self.value = self.from_json(json_str)
+
+class DataStreamReader(ABC):
+    def latest_offset(self) -> DataStreamOffset:
+        ...
+
+    def partitions(self, start: DataStreamOffset, end: DataStreamOffset) -> Sequence[InputPartition]:
+        ...
+
+    def read(self, partition) -> Iterator[Union[Tuple, Row]]:
+        ...
+
+    def offset_to_json(self, offset: DataStreamOffset) -> str:
+        ...
+
+    def json_to_offset(self, json: str) -> DataStreamOffset:
```

Comment:
I was trying to make this method a member of DataStreamOffset, but the json string doesn't contain any type information and can't be loaded back to original class(we enforce using plaintext json because @HeartSaVioR don't want to store unreadable pickled bytes in offset log). I wonder if @HyukjinKwon have any good suggestion about the interface of deserializing json back to specific DataStreamOffset as a Python expert.